doc/reference/trd-appid.md

```diff
@@ -0,0 +1,787 @@
+credentials of an application, they are typically large data
+structures that are too large to store in RAM. When parts of the
+kernel wish to apply security or access policies based on Application
+Identifiers, they need a concise way to represent these
+policies. Requiring policies to be encoded in terms of Application
+Identifiers is extremely costly: a table, for example, that says that
+only Applications signed with a particular 4096-bit RSA key can access
+certain system calls requires storing the whole 4096-bit key. If there
+are multiple such security policies through the kernel, they must each
+store this information.
+
+The `Compress` trait provides a mechanism to map an Application
+Identifier to a small (32-bit) integer called a Short ID. Short IDs
+can be used throughout the kernel as an identifier for security
+policies. 
+
+For example, suppose that a device wants to grant access to
+all Userspace Binaries signed by a certain 3072-bit RSA key K and has
+no other security policies. The Credentials Checking Policy only
+accepts `Rsa3072KeyWithID` credentials with key K. The `Compress`
+trait implementation assigns a Short ID as the ID stored in the
+credentials. Access control systems within the kernel can define their
+policies in terms of these identifiers, such that they can check
+access by comparing 32-bit integers rather than 384-byte keys.
+
+The 32-bit value MUST be non-zero. `ShortID` uses `core::num::NonZeroU32`
+so that an `Option<ShortID>` can be 32 bits in size, with 0 reserved
+for `None`.
+
+```rust
+#[derive(Clone, Copy, Eq)]
+struct ShortID {
+    id: core::num::NonZeroU32
+}
+
+pub trait Compress {
+    fn to_short_id(process: &dyn Process) -> Option<ShortID>;
+}
+```
+
+The `to_short_id` method returns an `Option` so that it has a clear
+default action if the Process Checker does not wish to assign it any
+meaningful identifier that might allow additional access.  A return
+value of `None` semantically means that the process's Application
+Credentials or other state do not map to any known security group or
+set of privileges. A`Some` result means the credentials map to a known
+security group or set of privileges.
+
+Generally, the Process Checker that implements `AppCredentialsChecker`
+and `AppIdentification` also implements `Compress`. This allows it to
+share copies of public keys or other credentials that it uses to make
+decisions, reducing flash space dedicated to these constants. Doing so
+also makes it less likely that the two are inconsistent.
+ 
+`ShortID` values MUST be locally unique among running processes.  The
+mapping between Global Application Identifiers and `ShortID` values
+MUST be deterministic.  Kernels SHOULD implement `Compress` in a
+manner that minimizes the chance that two different Application
+Identifiers compress to the same Short ID (e.g., taking the low-order
+bits of a strong cryptographic hash function, or using a known,
```

Comment:
Ah, I didn't see the use case before but now I see it.

For future reference, one use case is if a Tock system wants both of the following features:

1. Over-the-air app updates
1. Secure boot (i.e. the kernel has a fixed list of app IDs and it will only execute apps in the list)

Then the apps must be signed by public keys (to enable the OTA updates) and you can trust that the public keys themselves are not malicious (because of the secure boot). That combination allows you to use the low-order bits of the hash of the public key as the short ID.